alg/sha256_sse2.c

```diff
@@ -189,17 +185,18 @@ SHA256_Transform_sse2(uint32_t state[static restrict 8],
     uint32_t S[static restrict 8])
 #endif
 {
+	__m128i Y[4];
 	int i;
 
 	/* 1. Prepare the first part of the message schedule W. */
-	_mm_storeu_si128((__m128i *)&W[0],
-	    _mm_bswap_epi32(_mm_loadu_si128((const __m128i *)&block[0])));
-	_mm_storeu_si128((__m128i *)&W[4],
-	    _mm_bswap_epi32(_mm_loadu_si128((const __m128i *)&block[16])));
-	_mm_storeu_si128((__m128i *)&W[8],
-	    _mm_bswap_epi32(_mm_loadu_si128((const __m128i *)&block[32])));
-	_mm_storeu_si128((__m128i *)&W[12],
-	    _mm_bswap_epi32(_mm_loadu_si128((const __m128i *)&block[48])));
+	Y[0] = _mm_bswap_epi32(_mm_loadu_si128((const __m128i *)&block[0]));
+	_mm_storeu_si128((__m128i *)&W[0], Y[0]);
```

Comment:
Admittedly, although MSG4 looks nicer, this code is a bit ugly.  Unfortunately, we face a slight performance penalty (-0.5% to -1%) if we don't store each `Y` value as soon as we calculate it.  :(

An alternate idea would be to add a function like this (untested):
```
static void
load_block(W, i, block)
{
    _mm128i a;

    a = _mm_loadu_si128((const __m128i *)&block[16*i]);
    a = _mm_bswap_epi32(a);
    _mm_storeu_si128((__m128i *)&W[4*i], a);
    return (a)
}
```

but that felt like it was being over-engineered.